@google-cloud/storage definitions

blove opened the pull request:
- [x] Make your PR against the master branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the readme.
- [x] Avoid common mistakes.
- [x] Run tsc without errors.
- [x] Run npm run lint package-name if a tslint.json is present.
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [NO] Create it with `npm run new-package package-name`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

NOTE: I was not able to run `npm run new-package` without getting an exception: `Unexpected crash! Please log a bug with the commandline you specified.
/Users/brian/open-source/DefinitelyTyped/node_modules/dts-gen/bin/lib/run.js:109
        throw e;
`

blove commented:
Please note, I have only completed the definitions for the [at]google-cloud/storage files:
- bucket.js
- channel.js
- file.js
- storage.js

I did create the root [at]google-cloud folder with a root index.d.ts definition file. Please let me know if I should rename this or structure it differently.

I hope this is a good start to creating definition files for the Google Cloud Node.js library.

yuit commented on @google-cloud/storage/bucket/index.d.ts:
```diff
@@ -0,0 +1,102 @@
+  UploadOptions
+} from "../storage";
+
+declare module "@google-cloud/storage/bucket" {
```
@blove Question: why you declare ambient module here instead of just declare "export class Bucket ..."etc at the top level? Is it that the library is meant to use with "require" as well

blove replied:
@yuit Honestly, I was just following the contribution guidelines for DT. I can change this to simply export the class declarations and interfaces.

yuit replied:
My comment about "Require" is not correct. Although you can use ambient module declaration with es6 style import, It will be better to pull these off to top levels.

blove replied:
Ok - done.

yuit commented:
@blove sorry for delay here...just have a question for you since I am not familiar with the library but so far looks good to me

andy-ms commented:
Let's not merge this until we have a decision on Microsoft/types-publisher#155.

yuit commented:
@blove thanks for the PR.  Please be patient with us while we try to resolve the scope package and will let you know what need to be done/not done. :)

blove commented:
@yuit and @andy-ms I'm happy to wait and help in any way that I can. I did rename the package from _[at]google-cloud_ to simply _google-cloud_ as suggested. I also removed the modules to expose the API at the top level.

paulvanbrenk commented:
Added a label, so I can easily keep track of progress here.

andy-ms commented:
This can be named `google-cloud__storage` now. (Microsoft/TypeScript#14819)
Note: you will have to use `declare module "@google-cloud/storage"` as we don't yet have actual compiler support for resolving scoped packages to types.

andy-ms commented:
We've recently updated the repository structure.
All packages are now in a `types` subdirectory.
Please rebase this pull request onto the latest `master` branch.

blove commented:
@andy-ms I rebased from master, moved the definitions <code>google-cloud__storage</code> into the new types directory and used the <code>declare module</code> ambient module syntax. I hope I did it correctly. 👍

andy-ms commented on types/google-cloud__storage/index.d.ts:
```diff
@@ -0,0 +1,376 @@
+
+  /**
+   * Bucket metadata.
+   * @interface BucketMetadata
```
Redundant JSDoc

andy-ms commented on types/google-cloud__storage/README.md:
```diff
@@ -0,0 +1,15 @@
+# Installation
```
Currently we won't publish a custom README. You could put anything important at the top of `index.d.ts`.

blove commented:
@andy-ms I removed the redundant JSDoc comments as well as the README file.

RyanCavanaugh commented:
Restarting Travis build

blove commented:
Thanks @RyanCavanaugh. 👍

RyanCavanaugh commented:
Lots of errors in there - please address. Thanks!

bogacg commented:
I'd like to use Google Cloud Storage more in apps I'm working on but not having types makes things harder and I ditch it honestly for the moment.

mhegazy commented:
@blove are you still interested in this PR? if so can you please address the CI test failures.

blove commented:
Thanks @mhegazy!
